### DISCUSSION TITLE
Maintaining up to date runtimes and nuget packages for APSIMx

### DISCUSSION THREAD

**Discussion by lie112** [APSIMInitiative/ApsimX#9023] [General]
Anyone got any ideas about a good workflow for keeping .Net runtimes up to date?
It seems like after every recent organisational forced Windows upgrade I am faced with a pop-up to upgrade the .Net Desktop Runtime when running APSIM from Visual Studio 2022. It seems to be a complicated process that usually takes many attempts. The latest request to upgrade to 6.0.31 will not work even if I have upgraded the full .Net SDK to 6.0.423, let alone the 6.0.31 Desktop runtime. Now it seems to be working backwards asking for 6.0.30!
There appear to be a number of depreciated Nuget packages included in the APSIMx solution for Azure storage and toolkit.parsers. Could these be the reason that I cannot get past the popup to upgrade? 
Is the upgrade of Nuget packages the responsibility of the solution or the development environment?

**Comment by lie112** [comment]
UPDATE: after a heap of installs, restarts, rebuilds, and full clean and rebuild, it finally came to life and runs.
Still worth discussing what Nuget packages we need in the list and what should be brought up to the latest versions.

**Reply by zur003** [reply]
Updating this stuff is a necessary nuisance. It's challenging it part because we have a lot of developers who may find they need to upgrade bits and pieces to keep in sync with the release build system. The fact that it's all a nuisance means that it tends to be put off until necessary, but updates will be necessary soon. We're currently targeting .NET 6 in our builds, but official support for .NET 6 by Microsoft will cease in November this year (that doesn't mean things would stop working, but there would be no more security and other updates). We ought to be migrating now to .NET 8, which will entail, among other things,, updating the various Nuget packages. Other issues like #8680 will need to be dealt with along the way, preferably sometime within the next few months. Sometimes you need to run pretty fast just to stay at the same place.

**Comment by zur003** [comment]
We will soon be due for another framework update, as support for .Net 8 will expire in November 2026. The good news is that moving to .Net 10 should be nearly painless. We have already updated (most of) the NuGet packages (via PR #10893). I can now target .Net 10 in a build with only these minor changes: (1) removing references to 2 NuGet packages that we're no longer actually using and (2) adding some trivial checks on the number of bytes read from a stream. We can move to .Net 10 whenever it is convenient.

The only small hiccup is an issue with updating the GtkSharp related NuGet packages, and that is only because the OxyPlot.GtkSharp3 package hasn't been updated in several years. @hol430 has raised this as an issue on their github repository, but there has been no response. If we want to use the updated GtkSharp packages (we don't really need to; the old ones still work with .Net 10), we may want to fork the OxyPlot,GtkSharp3 stuff and maintain it ourselves; the needed changes are trivial.

**Reply by hol430** [reply]
Yeah, for now it's easy to build your own OxyPlot.GtkSharp3 with updated GtkSharp dependencies (although good luck getting the gtk2 or .net framework gtk3 versions to build). Even updating it to run on top of gtk4 isn't too hard. However, the main OxyPlot project doesn't seem to be actively developed so I wonder if a better long term solution would involve looking into alternative plotting libraries. Of course, one would likely need to write a gtk wrapper mostly from scratch...

**Reply by par456** [reply]
The graphing system in general has become a scope creeped mess of features and patches, it will need a refactor at some point in the future, it's just going to be a massive job and for the near future we need to devote time back into bugfixes and smaller features again for a bit.

I'm pretty happy just kicking along the current OxyPlot.GtkSharp3 package while it works, but I think we'll need to look into a new graphing library once that refactor happens. 

That kind of fits into the general discussion around where we are going in regards to the GUI and gtk in the future, and how we might be able to slowly migrate away from gtk libraries over time. Dean started some of that with the Table systems, and maybe Graphs will have to be done in a similar way.
